content/6_Advanced/Lagrange.mdx

```diff
@@ -27,6 +23,144 @@ adding lambda\*smth
+
+<Spoiler title="Dynamic Programming Solution">
+Let's have $\texttt{dp}[i][j:\{0,1\}]$ represent the maximum sum if we consider the first $i$ elements, given that $j=0/1$ implies whether
+element $i$ is part of a subarray. Let $\texttt{cnt}[i][j]$ represent the number of
+people used in an optimal arrangement of $\texttt{dp}[i][j]$.
+
+For our $\texttt{dp}$ transitions, we have
+
+$$
+\{\texttt{dp}[i][0], \texttt{cnt}[i][0]\} =
+\max\begin{cases} \{\texttt{dp}[i - 1][0], \texttt{cnt}[i - 1][0]\}\\
+\{\texttt{dp}[i - 1][1], \texttt{cnt}[i - 1][1]\}
+\end{cases}
+$$
+
+and
+
+$$
+\{\texttt{dp}[i][1], \texttt{cnt}[i][1]\} =
+\max\begin{cases}\{\texttt{dp}[i - 1][0] + A[i] - \lambda, \texttt{cnt}[i - 1][0] + 1\}\\
+\{\texttt{dp}[i - 1][1] + A[i], \texttt{cnt}[i - 1][1]\}\end{cases}
+$$
+
+because we either begin a new subarray or we continue an existing subarray.
+</Spoiler>
+
+Let $v$ be the maximal achievable sum with $\lambda$ penalty and $c$ be the number of subarrays used to achieve $v$. Then the **maximal possible sum achievable if we use exactly $c$ subarrays is $v+\lambda c$**. Note that we add $\lambda c$ to undo the penalty.
+
+Our goal is to find some $\lambda$ such that $c=K$ (assuming $K$ is at most the number of positive elements). As we increase $\lambda$, it makes sense for $c$ to decrease since we are penalizing subarrays more. Thus, we can try to binary search for $\lambda$ to make $c=K$ and set our answer to be $v+\lambda c$ at the optimal $\lambda$.
+
+This idea almost works but there are still some very important caveats and conditions that we have not considered.
+
+### Geometry
+
+Let $f(x)$ be the maximal sum if we use at most $x$ subarrays. We want to find $f(K)$.
+
+The first condition is that $f(x)$ **must be concave or convex**. Since $f(x)$ is increasing in this problem, the means that we need $f(x)$ to be concave: $f(x) - f(x - 1) \ge f(x + 1) - f(x)$. In other words, this means that the more subarrays we add, the less we increase the sum by. We can intuitively see that this is true.
+
+<Spoiler title="Proof that our function is concave">
+
+We construct a flow graph with source $S$, sink $T$, and $N+1$ additional vertices numbered $1$ to $N+1$. We will have the following edges.
+
+- A directed edge from $S$ to $i$ ($1 \le i \le N+1$) with weight $0$ and capacity $1$.
+
+- A directed edge from $i$ ($1 \le i \le N+1$) to $T$ with weight $0$ and capacity $1$.
+
+- A bidirectional edge from $i$ ($1 \le i \le N$) to $i+1$ with weight $A[i]$ and capacity $1$.
+
+$f(x)$ will be the maximum cost $x$-flow through the graph. We can repeatidly find the maximum cost augmenting path $x$ times to get our answer. Because the maximum cost as a function of flow is concave, $f(x)$ will be concave. You can read more about simulating cost flows [here](https://codeforces.com/blog/entry/118391).
```

Comment:
```suggestion
$f(x)$ will be the maximum cost $x$-flow through the graph. We can repeatedly find the maximum cost augmenting path $x$ times to get our answer. Because the maximum cost as a function of flow is concave, $f(x)$ will be concave. You can read more about simulating cost flows [here](https://codeforces.com/blog/entry/118391).
```